src/topology/category/Compactum.lean

```diff
@@ -0,0 +1,467 @@
+  apply le_nhds_of_str_eq,
+  rw [←str_hom_commute, str_eq_of_le_nhds ⟨_, h1⟩ x h2],
+end
+
+/-- Given any compact Hausdorff space, we construct a Compactum. -/
+noncomputable def of_topological_space (X : Type*) [topological_space X]
+  [compact_space X] [t2_space X] : Compactum :=
+{ A := X,
+  a := ultrafilter.Lim,
+  unit' := by {ext x, exact Lim_eq (by finish [le_nhds_iff]) },
+  assoc' := begin
+    ext FF,
+    change ultrafilter (ultrafilter X) at FF,
+    set x := (ultrafilter.map ultrafilter.Lim FF).Lim with c1,
+    have c2 : ∀ (U : set X) (F : ultrafilter X), F.Lim ∈ U → is_open U → U ∈ F.1,
+    { intros U F h1 hU,
+      exact c1 ▸ is_open_iff_ultrafilter.mp hU _ h1 _ F.2 (is_ultrafilter.le_nhds_Lim _) },
+    have c3 : (ultrafilter.map ultrafilter.Lim FF).1 ≤ nhds x,
+    { rw le_nhds_iff,
+      intros U hx hU,
+      refine c2 _ _ (by rwa ← c1) hU },
+    have c4 : ∀ (U : set X), x ∈ U → is_open U → { G : ultrafilter X | U ∈ G.1 } ∈ FF.1,
+    { intros U hx hU,
+      suffices : ultrafilter.Lim ⁻¹' U ∈ FF.1,
+      { apply filter.sets_of_superset _ this,
+        intros P hP,
+        exact c2 U P hP hU},
+      exact @c3 U (mem_nhds_sets hU hx) },
+    apply Lim_eq,
+    rw le_nhds_iff,
+    exact c4,
+  end }
+
+/-- Any continuous map between Compacta is a morphism of compacta. -/
+def hom_of_continuous {X Y : Compactum} (f : X → Y) (cont : continuous f) : X ⟶ Y :=
+{ f := f,
+  h' := begin
+    rw continuous_iff_ultrafilter at cont,
+    ext (F : ultrafilter X),
+    specialize cont (X.str F) F.1 F.2 (le_nhds_of_str_eq F (X.str F) rfl),
+    have := str_eq_of_le_nhds (ultrafilter.map f F) _ cont,
+    simpa only [←this, types_comp_apply, of_type_functor_map],
+  end }
+
+end Compactum
+
+/-- The type of Compact Hausdorff topological spaces. -/
+structure CompHaus :=
+(to_Top : Top)
+[is_compact : compact_space to_Top]
+[is_hausdorff : t2_space to_Top]
+
+namespace CompHaus
+
+instance : inhabited CompHaus := ⟨{to_Top := { α := pempty }}⟩
+
+instance : has_coe_to_sort CompHaus := ⟨Type*, λ X, X.to_Top⟩
+instance {X : CompHaus} : compact_space X := X.is_compact
+instance {X : CompHaus} : t2_space X := X.is_hausdorff
+
+instance category : category CompHaus := induced_category.category to_Top
+
+end CompHaus
+
+/-- The functor functor from Compactum to CompHaus. -/
+def Compactum_to_CompHaus : Compactum ⥤ CompHaus :=
+{ obj := λ X, { to_Top := { α := X } },
+  map := λ X Y f,
+  { to_fun := f,
+    continuous_to_fun := Compactum.continuous_of_hom _ }}
+
+namespace Compactum_to_CompHaus
+
+/-- The functor Compactum_to_CompHaus is full. -/
+def full : full Compactum_to_CompHaus :=
+{ preimage := λ X Y f, Compactum.hom_of_continuous f.1 f.2 }
+
+/-- The functor Compactum_to_CompHaus is faithful. -/
+lemma faithful : faithful Compactum_to_CompHaus := {}
+
+private lemma is_open_iff_ultrafilter' {X : Type*} [topological_space X] [compact_space X]
+  [t2_space X] (U : set X) : is_open U ↔ (∀ F : ultrafilter X, F.Lim ∈ U → U ∈ F.1) :=
```

Comment:
Does this need any of the machinery from this file? If no then it should move to the ultrafilter file.

Reply:
I moved this lemma to the file `topology/separation.lean` right after `is_ultrafilter.Lim_eq_iff_le_nhds`. I think this location makes sense since `is_ultrafilter.Lim_eq_iff_le_nhds` is another lemma which uses both `t2_space` and `compact_space`.